### PR TITLE
[4.0] Components dashboard

### PR DESCRIPTION
**Pull request by brianteeman**
PR for #27770

Before this PR In the dashboards of joomla 4 we have a situation where sometimes there is an icon and sometimes not.

For all the static preset dashboards the icon was coded in the preset
For the dynamic dashboard for components there was no icon

With this PR then the components module will display icons now. For core components this is done by updating rhe menu table.

Any extension that already defines an icon eg com_pathctester will also work as this is fully backwards compatible

Extension developers can still add either an image or an icon via their components xml

eg
`		<menu img="class:icon_name">COM_MYCOMPONENT</menu>
`or
`		<menu img="pathto/image_name">COM_MYCOMPONENT</menu>
`or
`		<menu img="image:pathto/image_name">COM_MYCOMPONENT</menu>
`

~_This is a DRAFT. It is ready for testing but I haven't done the install sql or updated the core component xml yet until after feedback._~

### Documentation Changes Required

Hopefully none as the menu aspect of the xml should already be documented

### CODE REVIEW

**Comment by Bakual** [comment]
Looks like a good approach to me.
I guess the hardcoded stuff in the menu code can then be removed as well?

**Comment by brianteeman** [comment]
> I guess the hardcoded stuff in the menu code can then be removed as well?

Actually not as those components do not have an entry in the _menus table

**Comment by Bakual** [comment]
Aww, crap. That's true.
Maybe taking it from the menu preset could work? As I understood those presets there is one for the menu and then several for the dashboards. But I may be wrong, I still don't understand those presets.

**Comment by brianteeman** [comment]
Let's tackle one thing at a time ;)

If you could be so kind as to test what I have done so far then I can finish it
- sql installs
- postgresql updates
- component xml (not used but would be good to be consistent)

Note one thing about my implementation is that it does allow for dashboards to be created when using the alternate template as it uses the same field and formatting for the icon/image (assuming someone creates the dashboard code.

There were many ways to achieve this but I really tried for the most compatible method - not necessarily the cleverest.

File: administrator/modules/mod_submenu/tmpl/default.php [hunk 1]
```diff
@@ -16,10 +16,29 @@
 /** @var  \Joomla\CMS\Menu\MenuItem  $root */
 ?>
 <?php foreach ($root->getChildren() as $child) : ?>
+
 	<?php if ($child->hasChildren()) : ?>
 		<div class="card">
+			<?php
+			if (substr($child->img, 0, 6) === 'class:')
```
Comment (Quy) [inline administrator/modules/mod_submenu/tmpl/default.php]: Indent lines 23, 27, 31, 35

**Comment by Bakual** [comment]
I have tested this item :white_check_mark: successfully on 9b2a716bb66dfd581ee155e0d1514f7e8644bc37<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27773">issues.joomla.org/tracker/joomla-cms/27773</a>.</sub>

**Comment by Bakual** [comment]
I even tried adding ` img="class:comment"` to my component menu declaration and it now shows an icon for it in the "Components Dashboard".
So that imho is a great solution which also works for 3rd parties.

**Comment by brianteeman** [comment]
Thanks @bakual - I will complete the PR now

**Comment by brianteeman** [comment]
No longer a draft and ready for testing

**Comment by Bakual** [comment]
Just noticed that News Feeds doesn't show the RSS icon you specified. I guess that is no longer (?) part of our FA package?
Or I did something wrong with all the composer/npm stuff

**Comment by brianteeman** [comment]
seems ok to me
https://fontawesome.com/icons/rss?style=solid

**Comment by Bakual** [comment]
It may well be that I messed up something in my local installation. If it shows for you and others, then it's fine.

File: administrator/modules/mod_submenu/tmpl/default.php (hunk 1 above)
Comment (Quy) [inline administrator/modules/mod_submenu/tmpl/default.php]: Indent lines 23, 27, 31, 35.

Comment (brianteeman) [reply administrator/modules/mod_submenu/tmpl/default.php]: I will do it this time as one of my last commits but please note that the rest of this file does not follow this indent after php

File: administrator/modules/mod_submenu/tmpl/default.php [hunk 2]
```diff
@@ -16,10 +16,29 @@
 /** @var  \Joomla\CMS\Menu\MenuItem  $root */
 ?>
 <?php foreach ($root->getChildren() as $child) : ?>
+
 	<?php if ($child->hasChildren()) : ?>
 		<div class="card">
+			<?php
+				if (substr($child->img, 0, 6) === 'class:')
+					{
```
Comment (Quy) [inline administrator/modules/mod_submenu/tmpl/default.php:24]: `{}` should align with `if` and `elseif`

File: administrator/modules/mod_submenu/tmpl/default.php [hunk 3]
```diff
@@ -16,10 +16,29 @@
 /** @var  \Joomla\CMS\Menu\MenuItem  $root */
 ?>
 <?php foreach ($root->getChildren() as $child) : ?>
+
```
Comment (Quy) [inline administrator/modules/mod_submenu/tmpl/default.php]:
```suggestion
```

File: administrator/modules/mod_submenu/tmpl/default.php [hunk 4]
```diff
@@ -18,8 +18,26 @@
...
+					$iconImage = '<img src="' . substr($child->img, 6) . '" aria-hidden="true">';
+				}
+				elseif (!empty($child->img))
+				{
+					$iconImage = '<img src="' . $child->img . '" aria-hidden="true">';
+				}
+				elseif ($child->icon)
+				{
+					$iconImage = '<span class="fas fa-' . $child->icon . '" aria-hidden="true"></span>';
+				}
```
Comment (Quy) [inline administrator/modules/mod_submenu/tmpl/default.php]: Would it be possible for a component to not have an image/icon that would result in `Undefined variable: iconImage` or being assigned the previous image/icon?

Comment (brianteeman) [reply administrator/modules/mod_submenu/tmpl/default.php]: I dont k now you would need to test

Comment (Quy) [reply administrator/modules/mod_submenu/tmpl/default.php]: To be on the safe side, add an `else` after line 37 to be blank or assign a default icon.

**Comment by infograf768** [comment]
Could you post a screenshot of the resulting modification? For example concerning Multilingual Associations.

**Comment by brianteeman** [comment]
I am not aware of any way this impacts on multi lingual associations. But try it and see.

**Comment by infograf768** [comment]
It impacts it by adding an icon in the dashboard.
For testers to know what to expect here is a screenshot after patch where the icons are present next to each component title.
Screen Shot 2020-02-08 at 11 30 05

**Comment by infograf768** [comment]
conflicts

**Comment by brianteeman** [comment]
It's not an impact it is the desired and intended behaviour as described in the pull request

**Comment by infograf768** [comment]
Cool down. By impact, i meant modify and it looks good. (I am not British).
conflicts remain to be solved.

**Comment by brianteeman** [comment]
conflicts resolved

File: administrator/modules/mod_submenu/tmpl/default.php [hunk 5]
```diff
@@ -17,39 +17,57 @@
...
+				else $iconImage = '';
```
Comment (Quy) [inline administrator/modules/mod_submenu/tmpl/default.php]:
```suggestion
				else
				{
					$iconImage = '';
				}
```

**Comment by Bakual** [comment]
@brianteeman I wanted to test your last changes again, and got a very messed up page. To verify it's not on my end, I've did a complet fresh installation and it was still the same. So probably a wrong conflict solving. Can you check if that's the same for you?
"News Feeds" is still not showing the icon - don't know what's wrong here and why it's showing for you and JM (according to his screenshot).
image

**Comment by brianteeman** [comment]
probably a conflict with the changes in #27777

**Comment by Bakual** [comment]
Indeed, with this PR you're reverting some of the changes you did there. Eg the `<div class="module-wrapper">` is removed again. Probably other changes as well. You need to check that layout again.

**Comment by sakiss** [comment]
It breaks the layout in my case:
component_dash_before
After patch
component_dash_after

**Comment by brianteeman** [comment]
@bakual @sakiss Please can you retest. There was an extra closing div after the merge conflict fix that was breaking the display

**Comment by Quy** [comment]
Still display issue on a new installation.

**Comment by Bakual** [comment]
Still doesn't look right to me.
Have a look at administrator/modules/mod_submenu/tmpl/default.php. If I remember right, you only changed the lines regarding the icon, but now there is a whole bunch of lines changed in the PR.
Still looks like a bad merge conflict solving.

**Comment by brianteeman** [comment]
Sorry for the bad merge conflict - now I really am sure it is correct

File: administrator/modules/mod_submenu/tmpl/default.php [hunk 6]
```diff
@@ -19,8 +19,26 @@
...
+						$iconImage = '<img src="' . substr($child->img, 6) . '" aria-hidden="true">';
+					}
+					elseif (!empty($child->img))
+					{
+						$iconImage = '<img src="' . $child->img . '" aria-hidden="true">';
+					}
+					elseif ($child->icon)
+					{
+						$iconImage = '<span class="fas fa-' . $child->icon . '" aria-hidden="true"></span>';
+					}
```
Comment (Quy) [inline administrator/modules/mod_submenu/tmpl/default.php:38]:
```suggestion
					}
					else
					{
						$iconImage = '';
					}
```

**Comment by Bakual** [comment]
I have tested this item :white_check_mark: successfully on 1684df9a55287622ea05a9e47da2a6a3be3bc0e0<br /><br />Works perfectly fine again!
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27773">issues.joomla.org/tracker/joomla-cms/27773</a>.</sub>

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2020-02-02.sql [hunk 7]
```diff
@@ -0,0 +1,6 @@
+UPDATE `#__menu` SET `img` = 'class:bookmark' WHERE `path` = 'Banners';
+UPDATE `#__menu` SET `img` = 'class:address-book' WHERE `path` = 'Contacts';
+UPDATE `#__menu` SET `img` = 'class:envelope' WHERE `path` = 'Messaging';
```
Comment (Quy) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2020-02-02.sql]: No longer exists per #27096

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 4aed3fff8bc28412ea2b1ec340b4140367e00470<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27773">issues.joomla.org/tracker/joomla-cms/27773</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27773">issues.joomla.org/tracker/joomla-cms/27773</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Comment by brianteeman** [comment]
thx

**Pull request merged**